### PR TITLE
feat(java): add Optional<T> serialization/deserialization

### LINKED ISSUE #984: Allow Optional<T> serialization/deserialization
**Is your feature request related to a problem? Please describe.**

The firestore serializer/deserializer doesn't support Optional\<T\> fields.

**Describe the solution you'd like**

I'd like the firestore serializer/deserializer support reading from Optional\<T\> and parsing responses to Optional\<T\> fields.

**Describe alternatives you've considered**

n/a

**Additional context**

n/a

### PR DESCRIPTION
**Pull request by jpJuni0r**
This PR adds support for Optional\<T\> serialization/deserialization for the firestore.

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-firestore/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #984 ☕️

### CODE REVIEW

**Review by ywmei-brt1 — COMMENTED** [review]
Thank you very much for your PR, I have left some questions for you in the PR, and looking forward to hearing back from you.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/MapperTest.java [hunk 1]
```diff
@@ -1299,6 +1304,24 @@ public void packageGetterSerialize() {
...
   }
 
+  @Test
+  public void serializeOptional() {
+    assertEquals("foo", serialize(Optional.of("foo")));
+  }
+
+  @Test
+  public void deserializeOptional() {
+    OptionalStringBean bean = deserialize("{'foo': 'bar'}", OptionalStringBean.class);
```
[the remark this answers is not in the document]
Comment (jpJuni0r) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/MapperTest.java:1314]: Hi and thanks a lot for taking a look at the code. You raise an interesting point which I overlooked at the first Iteration!

When receiving null in our JSON, I think it best makes sense to set the property to Optional.empty() and not to null. Also, there's no distinction between the first and the second case you described, as Optional.empty() and Optional.ofNullable(null) are equivalent.

I've adjusted my implementation and added a test case for the described behaviour.

[the remark this answers is not in the document]
Comment (ywmei-brt1) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/MapperTest.java:1314]: Thank you very much for your contribution. We are discussing the best way to handle some edge cases right now. 
Just want to check with you that you have any use case for serializing something like `Optional<Optional<String>>`?
Or you think we should just not support the nested Optional object's serialization:

```
  private static class OptionalInnerdBean{
    public Optional<String> bar;
  }
  private static class OptionalNestBean{
    public Optional<OptionalInnerdBean> foo;
  }

  @Test
  public void deserializeNestedOptional() {
    OptionalNestBean outer = new OptionalNestBean();
    OptionalInnerdBean inner = new OptionalInnerdBean();
    inner.bar = Optional.of("foo-bar");
    outer.foo = Optional.of(inner);
    
    assertExceptionContains(
            "throw some error message",
            () -> serialize(outer));
  }
  ```

[the remark this answers is not in the document]
Comment (jpJuni0r) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/MapperTest.java:1314]: I guess the best way to serialize null to `Optional<Optional<String>>` would be `Optional.of(Optional.empty())`. Would like to have any other modifications or other edge cases handled? Then I'll adjust the code accordingly and also add a test (incl. implementation) for the case you just described.

[the remark this answers is not in the document]
Comment (ywmei-brt1) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/MapperTest.java:1314]: Hi, thank you very much for the update. Apologize for getting back to you late. 

We had some meetings within the team, and decided we are going to support serialization of `Optional<T>` in a different way. There are a few reasons why we hesitate about direct supporting serialization of `Optional<T>`:

1. We need to keep our Java server SDK aligned with Android SDK (but this reflection requires android API level 29, which is too high for now );
2. Different customers might want to handle the serialization of `Optional` in different approach, we are leaning to provide a mechanism to customers so that they can build their own custom serializers in the future to handle these cases.

[the remark this answers is not in the document]
Comment (jpJuni0r) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/MapperTest.java:1314]: Okay, thanks a lot for the feedback! I guess I can close this PR now.
